File: kibana/scripts/upload_workflows.py
#!/usr/bin/env python3
# Upserts Kibana workflows from a directory of YAML files.
#
# Required env vars:
#   WORKFLOWS_DIR   — path to directory containing *.yml / *.yaml workflow files
#
# Optional env vars (read from kibana/.env if present):
#   KIBANA_URL      — default: http://localhost:5601
#   KIBANA_USERNAME — default: elastic
#   KIBANA_PASSWORD — default: changeme
import base64, json, os, sys, urllib.error, urllib.request
from pathlib import Path

# ── Config ────────────────────────────────────────────────────────────────────

def load_dotenv(path):
    if not path.exists():
        return
    for line in path.read_text().splitlines():
        line = line.strip()
        if not line or line.startswith('#') or '=' not in line:
            continue
        key, _, val = line.partition('=')
        key = key.strip()
        if key and key not in os.environ:
            os.environ[key] = val.strip()

KIBANA_DIR = Path(__file__).resolve().parent.parent
load_dotenv(KIBANA_DIR / '.env')

WORKFLOWS_DIR = os.environ.get('WORKFLOWS_DIR', '')
if not WORKFLOWS_DIR:
    print("No WORKFLOWS_DIR set. Skipping.")
    sys.exit(0)

KIBANA_URL      = os.environ.get('KIBANA_URL',      'http://localhost:5601').rstrip('/')
KIBANA_USERNAME = os.environ.get('KIBANA_USERNAME', 'elastic')
KIBANA_PASSWORD = os.environ.get('KIBANA_PASSWORD', 'changeme')

_AUTH = base64.b64encode(f'{KIBANA_USERNAME}:{KIBANA_PASSWORD}'.encode()).decode()
_HEADERS = {
    'Authorization':            f'Basic {_AUTH}',
    'kbn-xsrf':                 'true',
    'x-elastic-internal-origin': 'Kibana',
    'Content-Type':             'application/json',
}

# ── HTTP helpers ───────────────────────────────────────────────────────────────

def api(method, path, body=None):
    data = json.dumps(body).encode() if body is not None else None
    req  = urllib.request.Request(KIBANA_URL + path, data=data, headers=_HEADERS, method=method)
    try:
        with urllib.request.urlopen(req) as resp:
            return json.loads(resp.read())
    except urllib.error.HTTPError as e:
        raise RuntimeError(f'{method} {path} → HTTP {e.code}: {e.read().decode()}')

# ── Fetch existing workflows (paginated) ──────────────────────────────────────

def fetch_all_workflows():
    by_name, page = {}, 1
    while True:
        resp    = api('POST', '/api/workflows/search', {'size': 100, 'page': page, 'query': ''})
        results = resp.get('results', [])
        for w in results:
            by_name[w['name']] = w['id']
        if page >= resp.get('total', 1) or not results:
            break
        page += 1
    return by_name

def delete_if_exists(workflow_id):
    try:
        api('DELETE', f'/api/workflows/{workflow_id}')
    except RuntimeError as e:
        if 'HTTP 404' not in str(e):
            raise

def extract_field(lines, field):
    prefix = f'{field}:'
    return next(
        (l[len(prefix):].strip().strip('"\'') for l in lines if l.startswith(prefix)),
        None,
    )

def strip_id(raw_yaml):
    return '\n'.join(l for l in raw_yaml.splitlines() if not l.startswith('id:'))

# ── Main ──────────────────────────────────────────────────────────────────────

workflow_files = sorted(
    p for p in Path(WORKFLOWS_DIR).iterdir()
    if p.suffix in ('.yml', '.yaml')
)
if not workflow_files:
    print(f"No YAML files found in {WORKFLOWS_DIR}. Skipping.")
    sys.exit(0)

print(f"=== Fetching existing workflows from {KIBANA_URL} ===")
existing_by_name = fetch_all_workflows()
print(f"    Found {len(existing_by_name)} existing workflow(s)")

errors = 0
for path in workflow_files:
    raw_yaml = path.read_text()
    lines    = raw_yaml.splitlines()

    wf_id = extract_field(lines, 'id')
    name  = extract_field(lines, 'name')

    if not name:
        print(f"ERROR [{path.name}]: no top-level 'name:' field found", file=sys.stderr)
        errors += 1
        continue

    try:
        if wf_id:
            print(f"=== Replacing by id ({wf_id}): {name} ===")
            delete_if_exists(wf_id)
        elif name in existing_by_name:
            print(f"=== Replacing by name: {name} ===")
            api('DELETE', f'/api/workflows/{existing_by_name[name]}')
        else:
            print(f"=== Creating: {name} ===")

        result = api('POST', '/api/workflows', {'yaml': strip_id(raw_yaml)})
        print(f"    id: {result.get('id')}")
    except Exception as e:
        print(f"ERROR [{name}]: {e}", file=sys.stderr)
        errors += 1

print("=== Done ===")
sys.exit(1 if errors else 0)

File: kibana/scripts/upload_workflows.js
#!/usr/bin/env node
// Upserts Kibana workflows from a directory of YAML files.
//
// Required env vars:
//   WORKFLOWS_DIR   — path to directory containing *.yml / *.yaml workflow files
//
// Optional env vars (read from kibana/.env if present):
//   KIBANA_URL      — default: http://localhost:5601
//   KIBANA_USERNAME — default: elastic
//   KIBANA_PASSWORD — default: changeme
//
// Manifest:
//   If a manifest.yaml exists in WORKFLOWS_DIR, it controls per-workflow config.
//   Keys are workflow filenames (without extension). Supported options:
//     upload_count: N  — upload N copies of the workflow (default: 1)
//   Copies beyond the first get an ID/name suffix: "-2", "-3", etc.

const fs   = require('fs');
const path = require('path');
const yaml = require('js-yaml');

// ── Config ────────────────────────────────────────────────────────────────────

function loadDotenv(filePath) {
  if (!fs.existsSync(filePath)) return;
  for (const line of fs.readFileSync(filePath, 'utf8').split('\n')) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith('#') || !trimmed.includes('=')) continue;
    const [key, ...rest] = trimmed.split('=');
    const k = key.trim();
    if (k && !(k in process.env)) process.env[k] = rest.join('=').trim();
  }
}

const KIBANA_DIR = path.resolve(__dirname, '..');
loadDotenv(path.join(KIBANA_DIR, '.env'));

const WORKFLOWS_DIR = process.env.WORKFLOWS_DIR || '';
if (!WORKFLOWS_DIR) {
  console.log('No WORKFLOWS_DIR set. Skipping.');
  process.exit(0);
}

const KIBANA_URL      = (process.env.KIBANA_URL || 'http://localhost:5601').replace(/\/$/, '');
const KIBANA_USERNAME = process.env.KIBANA_USERNAME || 'elastic';
const KIBANA_PASSWORD = process.env.ES_PASSWORD || 'changeme';

const AUTH    = Buffer.from(`${KIBANA_USERNAME}:${KIBANA_PASSWORD}`).toString('base64');
const HEADERS = {
  'Authorization':       `Basic ${AUTH}`,
  'kbn-xsrf':            'true',
  'elastic-api-version': '2023-10-31',
  'Content-Type':        'application/json',
};

// ── Manifest ──────────────────────────────────────────────────────────────────

function loadManifest(dir) {
  const candidates = ['manifest.yaml', 'manifest.yml'];
  for (const name of candidates) {
    const p = path.join(dir, name);
    if (!fs.existsSync(p)) continue;
    const parsed = yaml.load(fs.readFileSync(p, 'utf8')) || {};
    const { ...workflows } = parsed;
    console.log(`    Loaded manifest: ${p}`);
    return { workflows };
  }
  console.log('    No manifest found — using defaults (upload_count: 1)');
  return { workflows: {} };
}

function getWorkflowConfig(manifest, fileBase) {
  const override = manifest.workflows[fileBase] || {};
  return { ...override };
}

// ── HTTP helpers ───────────────────────────────────────────────────────────────

async function kibanaApi(method, apiPath, body) {
  const res = await fetch(KIBANA_URL + apiPath, {
    method,
    headers: HEADERS,
    body: body !== undefined ? JSON.stringify(body) : undefined,
  });
  if (!res.ok) {
    const text = await res.text();
    throw new Error(`${method} ${apiPath} → HTTP ${res.status}: ${text}`);
  }
  const text = await res.text();
  return text ? JSON.parse(text) : null;
}

// ── Helpers ────────────────────────────────────────────────────────────────────

function copyName(name, index) {
  return index === 0 ? name : `${name} (${index + 1})`;
}

function copyId(fileBaseName, index) {
  return index === 0 ? fileBaseName : `${fileBaseName}-${index + 1}`;
}

// ── Upsert ─────────────────────────────────────────────────────────────────────
// PUT first (update). If 404 the workflow doesn't exist yet — POST to create.
// This avoids needing a list/search step entirely.

async function upsert(targetId, yamlForUpload) {
  try {
    await kibanaApi('PUT', `/api/workflows/workflow/${targetId}`, { yaml: yamlForUpload });
    console.log(`    Updated (${targetId})`);
  } catch (e) {
    if (!e.message.includes('HTTP 404')) throw e;
    await kibanaApi('POST', '/api/workflows/workflow', { yaml: yamlForUpload });
    console.log(`    Created (${targetId})`);
  }
}

// ── Concurrency ───────────────────────────────────────────────────────────────

async function runConcurrent(tasks, concurrency) {
  const active = new Set();
  for (const task of tasks) {
    const p = task().finally(() => active.delete(p));
    active.add(p);
    if (active.size >= concurrency) await Promise.race(active);
  }
  await Promise.all(active);
}

// ── Main ──────────────────────────────────────────────────────────────────────

async function main() {
  const workflowFiles = fs.readdirSync(WORKFLOWS_DIR)
    .filter(f => (f.endsWith('.yml') || f.endsWith('.yaml')) && f !== 'manifest.yaml' && f !== 'manifest.yml')
    .sort()
    .map(f => path.join(WORKFLOWS_DIR, f));

  if (!workflowFiles.length) {
    console.log(`No YAML files found in ${WORKFLOWS_DIR}. Skipping.`);
    process.exit(0);
  }

  console.log(`=== Loading manifest ===`);
  const manifest = loadManifest(WORKFLOWS_DIR);

  let errors = 0;
  const tasks = [];

  for (const filePath of workflowFiles) {
    const rawYaml = fs.readFileSync(filePath, "utf8");
    const fileBase   = path.basename(filePath, path.extname(filePath)).replace(/_/g, '-');
    const config     = getWorkflowConfig(manifest, fileBase);
    const uploadCount = Math.max(1, config.upload_count || 1);

    for (let i = 0; i < uploadCount; i++) {
      const targetId = copyId(fileBase, i);

      tasks.push(async () => {
        console.log(`=== Upserting: ${targetId} ===`);
        try {
          await upsert(targetId, rawYaml);
        } catch (e) {
          console.error(`ERROR [${targetId}]: ${e.message}`);
          errors++;
        }
      });
    }
  }

  await runConcurrent(tasks, 5);

  console.log('=== Done ===');
  process.exit(errors ? 1 : 0);
}

main().catch(e => { console.error(e); process.exit(1); });
